rasa/utils/tensorflow/layers.py

```diff
@@ -1137,11 +1146,36 @@ def _compute_softmax_loss(
                 [softmax_logits, sim_neg_ii, sim_neg_ll], axis=-1
             )
         # create label_ids for softmax
-        softmax_label_ids = tf.zeros_like(softmax_logits[..., 0], tf.int32)
-        softmax_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
-            labels=softmax_label_ids, logits=softmax_logits
+        # label_ids = tf.concat(
+        #     [
+        #         tf.ones_like(softmax_logits[..., :1], tf.float32),
+        #         tf.zeros_like(softmax_logits[..., 1:], tf.float32),
+        #     ],
+        #     axis=-1,
+        # )
+
+        # for numerical stability of softmax,
+        # see https://stackoverflow.com/a/49212689/3001665
+        input_logits = tf.nn.softmax(
+            softmax_logits - tf.math.reduce_max(softmax_logits, axis=-1, keepdims=True)
         )
-        return softmax_loss
+
+        loss = -tf.math.log(tf.math.maximum(input_logits[..., 0], 1e-6))
```

Comment:
Not sure what you tried here. We can...
1. plug in the softmax definition into the cross entropy, pull out the numerator, and let the log kill the one exp (nothing more that we can do there)
2. make use of shifting (same as in softmax) to get more stable results

which should yield something like
```suggestion
        logits = softmax_logits # because that name is just confusing
        shifted_logits = logits - tf.reduce_max(logits)
        cross_entropy = - logits[...,0] + tf.reduce_logsumexp(logits)
        loss = tf.reduce_sum( cross_entropy )
```
There's also`tf.nn.log_softmax` but that would give us more summands than we'd need.

Reply:
I assume in your suggestion `logits` is non-softmax applied version of logits. In that case, the above is exactly the same as my version.

Reply:
Nope, cause 1) there's no exp and then log taken from the numerator (which is what I'd think ce from logits would be doing), 2) there's no evaluation of softmax-normalised entires that aren't used, 3) don't think those changes would be optimized away

Reply:
> Nope

I meant mathematically its the same and the above version doesn't break on at least our datasets. I wanted to try it to test if our datasets can break it.

> cause 1) there's no exp and then log taken from the numerator

yeah that's reduced but the subraction of `tf.reduce_max(logits)` and `tf.maximum` would act as safeguards.
Surely we should take your version if we merge this.

> there's no evaluation of softmax-normalised entires that aren't used

Not sure what you mean

Reply:
It's not the same considering that we don't have infinite precision. What I mean by the unused entries is that tf.nn.softmax will return a tensor of the same shape as the input, while we only really need those entries corresponding to the positive examples.

Reply:
> It's not the same considering that we don't have infinite precision

Okay, you just mean that we should reduce the number of ops, so that precision is not lost in those ops at least, right?

> What I mean by the unused entries is that tf.nn.softmax will return a tensor of the same shape as the input, while we only really need those entries corresponding to the positive examples.

Sure thing, I didn't mean to have this as the most stable / polished version 👍